File: Dante1/Module1.04/Zadanie4.14/main.c
#include <stdio.h>

int is_non_monotonic(int a, int b, int c)
{
    return (a < b && b > c) || (a > b && b < c);
}

int main(void)
{
    int error;

    int array[4000];
    int size_of_array = 0;

    printf("Enter array: ");
    for(int i = 0; i < 4000; i++)
    {
        error = scanf("%i", &array[i]);
        if(error != 1)
        {
            printf("Incorrect input");
            return 1;
        }
        if(array[i] == 0)
        {
            break;
        }
        size_of_array++;
    }
    if(size_of_array > 1000)
    {
        size_of_array = 1000;
    }

    if (size_of_array < 3)
    {
        printf("0\n0\n");
        return 0;
    }

    printf("\n");

    int max_length = 0;
    int counter_for_biggest_subarray = 0;
    int helper_for_array[1000];

    for (int i = 0; i < size_of_array - 2; i++)
    {
        int current_length = 2;
        while (i + current_length < size_of_array && is_non_monotonic(array[i + current_length - 2], array[i + current_length - 1], array[i + current_length]))
        {
            current_length++;
        }
        if (current_length > max_length)
        {
            max_length = current_length;
            counter_for_biggest_subarray = 1;
            helper_for_array[0] = i;
        }
        else if (current_length == max_length)
        {
            helper_for_array[counter_for_biggest_subarray++] = i;
        }
        i += current_length - 2;
    }

    if (max_length >= 3)
    {
        printf("%i\n%i\n", max_length, counter_for_biggest_subarray);
        for (int i = 0; i < counter_for_biggest_subarray; i++)
        {
            printf("%i -", helper_for_array[i]);
            for (int j = 0; j < max_length; j++)
            {
                printf(" %i", array[helper_for_array[i] + j]);
            }
            printf("\n");
        }
    }
    else
    {
        printf("0\n0\n");
    }

    return 0;
}

File: Dante1/Module1.04/Zadanie4.14/main.rs
fn is_non_monotonic(a: i32, b: i32, c: i32) -> bool
{
    return (a < b && b > c) || (a > b && b < c);
}

fn main()
{
    let mut array:[i32; 4000] = [0; 4000];
    let mut size_of_array: usize = 0;
    
    println!("Enter: ");
    
    let mut input: String = String::new();
    for _i in 0 .. 4000
    {
        input.clear();
        std::io::stdin().read_line(&mut input).expect("Error");
        let number: i32 = input.trim().parse().expect("Error");
        
        if number == 0
        {
            break;
        }
        
        array[size_of_array] = number;
        size_of_array += 1;
    }
    
    if size_of_array > 1000
    {
        size_of_array = 1000;
    }
    
    if size_of_array < 3
    {
        println!("{}\n{}", 0, 0);
    }
    
    let mut max_length: i32 = 0;
    let mut counter_for_biggest_subarray: i32 = 0;
    
    let mut helper_array:[i32; 1000] = [0; 1000];
    
    for mut i in 0 .. size_of_array - 2
    {
        let mut helper: i32 = 2;
        
        while i as i32 + helper < size_of_array as i32 && 
            is_non_monotonic(array[i + helper as usize - 2], array[i + helper as usize - 1], array[i + helper as usize])
        {
            helper += 1;
        }
        
        if helper > max_length
        {
            max_length = helper;
            counter_for_biggest_subarray = 1;
            helper_array[0] = i as i32;
        }
        else if helper == max_length
        {
            helper_array[counter_for_biggest_subarray as usize] = i as i32;
            counter_for_biggest_subarray += 1;
        }
        
        i += helper as usize * 2;
    }
    
    if max_length >= 3
    {
        println!("{}\n{}", max_length, counter_for_biggest_subarray);
        
        for i  in 0 .. counter_for_biggest_subarray
        {
            print!("{} -", helper_array[i as usize]);
            for j in 0 .. max_length
            {
                print!(" {}", array[(helper_array[i as usize] + j) as usize]);
            }
            println!();
        }
    }
    else
    {
        println!("{}\n{}", 0 , 0);
    }
    
    return;
}
